fms/utils/serialization.py

```diff
@@ -136,12 +188,270 @@ def load_state_dict(
 
         checkpoints = [checkpoints[rank]]
 
-    checkpoint_sds = [
-        torch.load(ckpt_path, map_location="cpu") for ckpt_path in checkpoints
-    ]
+    if file_format == "st":
+        from safetensors import safe_open  # type: ignore
+
+        # For safetensors, delay loading the weights until information about the model
+        # is available for sharding
+        checkpoint_sds = []
+        for ckp in checkpoints:
+            with safe_open(
+                ckp, framework="pt", device=str(initial_device)
+            ) as ckp_f:  # type: ignore[attr-defined]
+                st_sd = {}
+                for key in ckp_f.keys():
+                    st_sd[key] = {"file": ckp, "orig_key": key}
+                checkpoint_sds.append(st_sd)
+    else:
+        checkpoint_sds = [
+            torch.load(ckpt_path, map_location="cpu") for ckpt_path in checkpoints
+        ]
     assert len(checkpoint_sds[0]), f"Unable to load checkpoint data at {model_path}"
     if len(checkpoint_sds) == 1:
-        return checkpoint_sds[0]
+        return checkpoint_sds[0], file_format
```

Comment:
i don't love the tuple return type, maybe we should make it two functions -- e.g. a function to decide the checkpoint format and then make it a required parameter?

Reply:
reworked the code